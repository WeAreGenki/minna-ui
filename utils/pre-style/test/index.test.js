/** @jest-environment node */

/* tslint:disable no-implicit-dependencies */

// TODO: Add test for returned `dependencies`

'use strict';

const fs = require('fs');
const { promisify } = require('util');
const { preprocess, compile } = require('svelte');
const postcssNested = require('postcss-nested');
const preprocessStyle = require('../index.js');

/* eslint-disable-next-line security/detect-non-literal-fs-filename */
const readFile = promisify(fs.readFile);

// don't require() component to avoid Jest transform
const componentPath = require.resolve('@minna-ui/jest-config/fixtures/TestComponent.html');

const preprocessOpts = {
  style: preprocessStyle({
    plugins: [postcssNested],
  }),
};

const sourceSimple = `
<div id="target">test</div>

<style type="text/postcss">
  #target {
    color: red;

    &:focus {
      background: #000;
    }
  }
</style>
`;
const sourceNoTypeAttr = `
<div id="target">test</div>

<style>
  #target {
    color: red;

    &:focus {
      background: #000;
    }
  }
</style>
`;
const sourceBadSyntax = `
<style type="text/postcss">
  #target {
    color: red;
</style>
`;
let source = '';

beforeAll(async () => {
  source = await readFile(componentPath, 'utf8');
});

describe('Svelte style preprocessor', () => {
  it('processes a simple component', async () => {
    expect.assertions(3);
    const output = preprocess(sourceSimple, preprocessOpts);
    await expect(output).resolves.toBeDefined();
    const result = (await output).toString();
    expect(result).not.toMatch('&:focus');
    expect(result).toMatchSnapshot();
  });

  it('adds a banner comment', async () => {
    expect.assertions(2);
    let result = await preprocess(sourceSimple, {
      ...preprocessOpts,
      banner: '/*!\n * minna-ui\n */\n',
    });
    result = result.toString();
    expect(result).not.toMatch('&:focus');
    expect(result).toMatchSnapshot();
  });

  it('does not process without type attribute', async () => {
    expect.assertions(2);
    let result = await preprocess(sourceNoTypeAttr, preprocessOpts);
    result = result.toString();
    expect(result).toMatch('&:focus');
    expect(result).toMatchSnapshot();
  });

  it('processes a component', async () => {
    expect.assertions(4);
    const output = preprocess(source, preprocessOpts);
    await expect(output).resolves.toBeDefined();
    const result = (await output).toString();
    expect(result).not.toBeFalsy();
    expect(result).not.toEqual('');
    expect(result).toMatchSnapshot();
  });

  it.skip("compiles a component's CSS", async () => {
    expect.assertions(2);
    const processed = await preprocess(source, preprocessOpts);
    const result = compile(processed.toString());
    expect(result.css.code).not.toBeFalsy();
    expect(result.css.code).toMatchSnapshot();
  });

  it('prints error on bad CSS syntax', async () => {
    expect.assertions(1);
    const spy = jest.spyOn(process.stderr, 'write');
    spy.mockImplementation(() => {}); // tslint:disable-line no-empty
    await preprocess(sourceBadSyntax, preprocessOpts);
    expect(spy).toHaveBeenCalled();
    spy.mockRestore();
  });
});
