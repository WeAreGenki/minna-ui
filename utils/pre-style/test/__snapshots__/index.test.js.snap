// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Svelte style preprocessor adds a banner comment 1`] = `
"
<div id=\\"target\\">test</div>

<style>
  #target {
    color: red;
  }

    #target:focus {
      background: #000;
    }
</style>
"
`;

exports[`Svelte style preprocessor compiles a component's CSS 1`] = `
":root{--body-margin:0}@media screen and (min-width: 30em){body{margin:var(--body-margin)
  }}.text.svelte-1qwrmfn{color:rgb(255, 0, 0)}.text.svelte-1qwrmfn:focus{background:rgb(0, 0, 0)}.wrapper.svelte-1qwrmfn>.text.svelte-1qwrmfn{margin-top:1rem}"
`;

exports[`Svelte style preprocessor does not process without type attribute 1`] = `
"
<div id=\\"target\\">test</div>

<style>
  #target {
    color: red;

    &:focus {
      background: #000;
    }
  }
</style>
"
`;

exports[`Svelte style preprocessor processes a component 1`] = `
"<!--
  TEST COMPONENT

  A simple Svelte component for use in tests.

  NOTE:
    In the code, object properties starting with _ are safe to mangle; the name
    can be shortened at build time for smaller JavaScript bundles.
-->

<div class=\\"wrapper\\">
  <div class=\\"text\\" ref:_target>test {_name}</div>

  <div class=\\"text\\" ref:_nameReversed>test <span>{_reversed}</span></div>
</div>

<script>
  export default {
    data: () => ({
      _name: 'Elon Musk',
    }),
    computed: {
      _reversed: ({ _name }) =>
        _name
          .split('')
          .reverse()
          .join(''),
    },
  };
</script>

<style>
  $body-margin: 0;

  /* stylelint-disable no-descending-specificity */
  @media screen and (min-width: 30em) {
  :global(body) {
      margin: $body-margin
  }
    }

  .text {
    color: rgb(255, 0, 0);
  }

  /* red */

  .text:focus {
      background: rgb(0, 0, 0); /* black */
    }

  .wrapper > .text {
      margin-top: 1rem;
    }
</style>
"
`;

exports[`Svelte style preprocessor processes a simple component 1`] = `
"
<div id=\\"target\\">test</div>

<style>
  #target {
    color: red;
  }

    #target:focus {
      background: #000;
    }
</style>
"
`;
